Adding ProgressUI to the UI

ganimtron-10 opened the pull request:
Creating a ProgressUI Widget which can be used as Loading bars.

https://user-images.githubusercontent.com/64432063/128671950-849c19b5-17ca-47a2-92a5-3f6e357538c5.mp4

https://user-images.githubusercontent.com/64432063/131996694-e7cf3e34-b010-499d-a7d7-832c2348859e.mp4

pep8speaks commented:
Hello @ganimtron-10! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`fury/ui/elements.py`](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/elements.py):

> [Line 5:80](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/elements.py#L5): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 3055:1](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/elements.py#L3055): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 3059:35](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/elements.py#L3059): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 3059:49](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/elements.py#L3059): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 3061:29](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/elements.py#L3061): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 3061:31](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/elements.py#L3061): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 3061:34](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/elements.py#L3061): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 3061:54](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/elements.py#L3061): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 3061:58](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/elements.py#L3061): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 3086:25](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/elements.py#L3086): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 3087:1](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/elements.py#L3087): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 3093:1](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/elements.py#L3093): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 3095:5](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/elements.py#L3095): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 3098:80](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/elements.py#L3098): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 3100:67](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/elements.py#L3100): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 3100:80](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/elements.py#L3100): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 79 characters)
> [Line 3102:56](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/elements.py#L3102): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 3103:80](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/elements.py#L3103): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 79 characters)

* In the file [`fury/ui/tests/test_elements.py`](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/tests/test_elements.py):

> [Line 791:1](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/tests/test_elements.py#L791): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 792:41](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/tests/test_elements.py#L792): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 793:33](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/tests/test_elements.py#L793): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 799:26](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/tests/test_elements.py#L799): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 805:23](https://github.com/fury-gl/fury/blob/eaef495f4a83e112ffff08322a27aaba231471fa/fury/ui/tests/test_elements.py#L805): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

ganimtron-10 committed:
adding progressui with tests and tutorials

ganimtron-10 committed:
adding tutorial file

ganimtron-10 commented:
Hi @skoudoro @antrikshmisri !!
Please review whether the logic for this UI seems correct or not.

antrikshmisri commented on docs/tutorials/02_ui/viz_progress.py:
```diff
@@ -0,0 +1,94 @@
+
+First, some imports.
+"""
+from fury import ui, window, actor
```
Could you update the import order.

antrikshmisri commented on fury/ui/elements.py:
```diff
@@ -3051,3 +3052,114 @@ def directory_click_callback(self, i_ren, _obj, listboxitem):
+
+        Parameters
+        ----------
+        size : (float, float)
```
`, optional`

antrikshmisri commented on fury/ui/elements.py:
```diff
@@ -3051,3 +3052,114 @@ def directory_click_callback(self, i_ren, _obj, listboxitem):
+        ----------
+        size : (float, float)
+            Center of the slider's center.
+        initial_value : float
```
`, optional`

antrikshmisri commented on fury/ui/elements.py:
```diff
@@ -3051,3 +3052,114 @@ def directory_click_callback(self, i_ren, _obj, listboxitem):
+            Center of the slider's center.
+        initial_value : float
+            Initial value of the slider.
+        min_value : float
```
`, optional`

antrikshmisri commented on fury/ui/elements.py:
```diff
@@ -3051,3 +3052,114 @@ def directory_click_callback(self, i_ren, _obj, listboxitem):
+            Initial value of the slider.
+        min_value : float
+            Minimum value of the slider.
+        max_value : float
```
`, optional`

antrikshmisri commented on fury/ui/elements.py:
```diff
@@ -3051,3 +3052,114 @@ def directory_click_callback(self, i_ren, _obj, listboxitem):
+            Minimum value of the slider.
+        max_value : float
+            Maximum value of the slider.
+        padding : float
```
`, optional`

antrikshmisri commented on fury/ui/elements.py:
```diff
@@ -3051,3 +3052,114 @@ def directory_click_callback(self, i_ren, _obj, listboxitem):
+            Maximum value of the slider.
+        padding : float
+            Distance between loader and background.
+        bg_color : (float, float, float)
```
`, optional`

antrikshmisri commented on fury/ui/elements.py:
```diff
@@ -3051,3 +3052,114 @@ def directory_click_callback(self, i_ren, _obj, listboxitem):
+            Distance between loader and background.
+        bg_color : (float, float, float)
+            Background color of progress bar.
+        progress_color : (float, float, float)
```
`, optional`

antrikshmisri reviewed (commented):
Hey @ganimtron-10, thanks for the PR, please see my comments below. I think the logic is fine, I just can't think of many asynchronous processes in FURY where a progress bar will be helpful. Can you mention some use cases where this might be helpful. Also, can you add a user hook which will be triggered when all the processes are complete for eg.  removing the progress bar from the scene or some other effect. Thanks!

ganimtron-10 committed:
updating docs and import order

ganimtron-10 commented:
Thank You @antrikshmisri for the review!!

> Can you mention some use cases where this might be helpful.

Below examples are my understanding about the process, I don't have a good idea about how it works!
- Recently "Creation of Molecules and Proteins" was under work so there if we change any single parameter while interacting with it, we would need to modify the whole molecule/protein again, while this time we could keep user hooked by showing the progress of the work completed using this.
- If in future Fury tries too implement transitioning between the scenes then there also we can use this.
- While we store some data into buffer before presenting it ie. online images or physics simulations data then also this can be used

I have also addressed all the comments and will add a hook as soon as possible.

ganimtron-10 committed:
adding on_complete hook

ganimtron-10 commented:
Hi @antrikshmisri !!
I added an on_complete hook which is called when the progressbar completes all the process. PTAL.
Thanks!

antrikshmisri commented on fury/ui/elements.py:
```diff
@@ -3051,3 +3052,118 @@ def directory_click_callback(self, i_ren, _obj, listboxitem):
+        self._ratio = self.progress.width/self.progress_total_size[0]
+
+        value_range = self.max_value - self.min_value
+        self._value = self.min_value + self.ratio * value_range
```
value is being set to a float instead of an integer. Hence, the value is never equal to its intended value.

antrikshmisri commented on fury/ui/elements.py:
```diff
@@ -3051,3 +3052,118 @@ def directory_click_callback(self, i_ren, _obj, listboxitem):
+
+    @value.setter
+    def value(self, value):
+        value_range = self.max_value - self.min_value
```
set `self._value` here instead of `update` and set it to the `value` parameter.

antrikshmisri commented on fury/ui/elements.py:
```diff
@@ -3051,3 +3052,118 @@ def directory_click_callback(self, i_ren, _obj, listboxitem):
+
+    def _get_actors(self):
+        """Get the actors composing this UI component."""
+        return self.background + self.loader
```
`self.background.actors` instead of `self.background`. Also, there is no attribute `loader`.

antrikshmisri reviewed (changes requested):
Hey @ganimtron-10 , thanks for the update. See my comments below. Your UI is missing the `resize` method. Also, can you demonstrate the use of `on_complete` hook in the tutorial.

ganimtron-10 committed:
updating doc strings

ganimtron-10 committed:
updating value setter, using on_complete hook in tutorials and updating tests

ganimtron-10 committed:
fix peps

ganimtron-10 committed:
fix peps

ganimtron-10 commented:
Hi @antrikshmisri , I have updated the code as per the comments and also added the `resize` method and `on_complete` hook tutorial.

ganimtron-10 committed:
upadting on_completing hook tutorial

ganimtron-10 committed:
Merge branch 'fury-gl:master' into progress-ui

antrikshmisri reviewed (commented):
Hey @ganimtron-10 , thanks for the update. Please see my comments below. Also, could you add a flag to make the UI draggable?

ganimtron-10 committed:
updating docs and some fixes

ganimtron-10 committed:
removing progressbar on completion

ganimtron-10 committed:
changing background to Panel2D

ganimtron-10 commented:
Hi @antrikshmisri ,
All the comments are been addressed and to make `UI` `draggable` and for more flexibility, I changed the background to `Panel2D`.
PTAL.
Thanks!

ganimtron-10 committed:
fix pep

ganimtron-10 committed:
updating fork form fury master

ganimtron-10 committed:
Merge branch 'master' of https://github.com/fury-gl/fury into fury-gl-master

ganimtron-10 committed:
git push origin progress-uiMerge branch 'fury-gl-master' into progress-ui

ganimtron-10 committed:
updating tests

ganimtron-10 committed:
Merge branch 'progress-ui' of https://github.com/ganimtron-10/fury into progress-ui
